src/pydvl/utils/parallel/futures/ray.py

```diff
@@ -0,0 +1,324 @@
+        ...
+        >>> print(results)
+        [1, 2, 3, 4, 5]
+    """
+
+    def __init__(
+        self,
+        max_workers: Optional[int] = None,
+        *,
+        config: ParallelConfig = ParallelConfig(),
+    ):
+        if config.backend != "ray":
+            raise ValueError(
+                f"Parallel backend must be set to 'ray' and not {config.backend}"
+            )
+        if max_workers is not None:
+            if max_workers <= 0:
+                raise ValueError("max_workers must be greater than 0")
+            max_workers = max_workers
+
+        config_dict = asdict(config)
+        config_dict.pop("backend")
+        config_dict.pop("n_workers")
+        if "address" not in config_dict:
+            config_dict["num_cpus"] = max_workers
+        self.config = config_dict
+        if not ray.is_initialized():
+            ray.init(**self.config)
+
+        self._max_workers = max_workers
+        if self._max_workers is None:
+            self._max_workers = int(ray._private.state.cluster_resources()["CPU"])
+
+        self._broken = False
+        self._shutdown = False
+        self._cancel_pending_futures = False
+        self._shutdown_lock = threading.Lock()
+        self._queue_lock = threading.Lock()
+        self._work_queue: "queue.Queue[Optional[_WorkItem]]" = queue.Queue(
+            maxsize=self._max_workers
+        )
+        self._pending_queue: "queue.SimpleQueue[Optional[_WorkItem]]" = (
+            queue.SimpleQueue()
+        )
+
+        # Work Item Manager Thread
+        self._work_item_manager_thread: Optional[_WorkItemManagerThread] = None
+
+    def submit(self, fn: Callable[..., T], *args, **kwargs) -> "Future[T]":
+        r"""Submits a callable to be executed with the given arguments.
+
+        Schedules the callable to be executed as fn(\*args, \**kwargs)
+        and returns a Future instance representing the execution of the callable.
+
+        :param fn: Callable.
+        :param args: Positional arguments that will be passed to `fn`.
+        :param kwargs: Keyword arguments that will be passed to `fn`.
+        :return: A Future representing the given call.
+        """
+        with self._shutdown_lock:
+            logger.debug("executor acquired shutdown lock")
+            if self._broken:
+                raise BrokenExecutor(self._broken)
+            if self._shutdown:
+                raise RuntimeError("cannot schedule new futures after shutdown")
+
+            logging.debug("Creating future and putting work item in work queue")
+            future: "Future[T]" = Future()
+            w = _WorkItem(future, fn, args, kwargs)
+            self._put_work_item_in_queue(w)
+            # We delay starting the thread until the first call to submit
+            self._start_work_item_manager_thread()
+            return future
+
+    def shutdown(self, wait: bool = True, *, cancel_futures: bool = False) -> None:
```

Comment:
Who calls this? How can one set `cancel_futures` to `True`?

Reply:
when using a concurrent.futures Executor class as a context manager, `shutdown()` is called automatically with the default arguments when exiting the block.

In order to call `shutdown()` and pass arguments to it we would have to instantiate the Executor normally and call shutdown manually.

Reply:
Yes, I understood what you meant. What I said in return was that we don't have to change the method signature, only the way we use the class. i.e. use

```python
executor = RayExecutor()
...
executor.shutdown(cancel_futures=True)
```

instead of

```python
with RayExecutor() as executor:
    ...
```

Reply:
😄 We are going in circles. I also understood you. But it is IMO an anti-pattern. When I say we should make it configurable, I mean:
```python
with RayExecutor(cancel_futures=True) as executor:
    ...
```
And I think that `True` should be the default, and cancel pending and already running futures as soon as the context is exited.

Reply:
I get your point but I would like to stick to the Executor's api. I found a solution that will satisfy us both: I changed the `__exit__()` method to set `cancel_futures=True` when calling the `shutdown()` method.

Reply:
As discussed during the meeting, I added a `cancel_futures_on_exit` parameter.